patches/api/0449-Add-Lifecycle-Event-system.patch

```diff
@@ -0,0 +1,577 @@
++     * lower numeric value correlates to the handler
++     * being run earlier.
++     *
++     * @param priority the numerical priority
++     * @return this configuration for chaining
++     */
++    @Contract("_ -> this")
++    PrioritizedLifecycleEventHandlerConfiguration<O> priority(int priority);
++
++    /**
++     * Sets this handler configuration to be considered a "monitor".
++     * These handlers will run last and should only be used by plugins
++     * to observe any changes from previously ran handlers.
++     *
++     * @return this configuration for chaining
++     */
++    @Contract("-> this")
++    PrioritizedLifecycleEventHandlerConfiguration<O> monitor();
++
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/Registrar.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/Registrar.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..fd9c3605a8f5e6bdd31e42f18a45154d4074eb67
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/Registrar.java
+@@ -0,0 +1,12 @@
++package io.papermc.paper.plugin.lifecycle.event.registrar;
++
++import org.jetbrains.annotations.ApiStatus;
++
++/**
++ * To be implemented by types that provide ways to register types
++ * either on server start or during a reload
++ */
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface Registrar {
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..04a3ea76a82c13abc21066cfe515b44627fc33af
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java
+@@ -0,0 +1,60 @@
++package io.papermc.paper.plugin.lifecycle.event.registrar;
++
++import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvent;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Contract;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * A lifecycle event that exposes a {@link Registrar} of some kind
++ * to allow management of various things. Look at implementations of
++ * {@link Registrar} for an idea of what uses this event.
++ *
++ * @param <R> registrar type
++ * @see Reloadable
++ */
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface RegistrarEvent<R extends Registrar> extends LifecycleEvent {
++
++    /**
++     * Get the registrar related to this event.
++     *
++     * @return the registrar
++     */
++    @Contract(pure = true)
++    @NotNull R registrar();
++
++    /**
++     * A lifecycle event that exposes a {@link Registrar} that is
++     * reloadable.
++     *
++     * @param <R>
++     * @see RegistrarEvent
++     */
++    @ApiStatus.Experimental
++    @ApiStatus.NonExtendable
++    interface Reloadable<R extends Registrar> extends RegistrarEvent<R> {
```

Comment:
Moving this out as `ReloadableRegistrarEvent` would make its use more intuitive I think